ENH: Multivariate t-distribution

gwgundersen opened the pull request:
### Reference issue

Closes https://github.com/scipy/scipy/issues/10042. Duplicates https://github.com/scipy/scipy/pull/11413 because the old branch was out-of-date.

### What does this implement/fix?

Adds support for the [multivariate t-distribution](https://en.wikipedia.org/wiki/Multivariate_t-distribution), particularly `pdf`, `logpdf`, and `rvs`. See [my blog post](http://gregorygundersen.com/blog/2020/01/20/multivariate-t/) for a high-level discussion of my approach.

### Additional information

**Completed:**

- [x] No licensing issues.
- [x] Unit tests comparing `logpdf` and `pdf` to MATLAB's `mvtpdf` function; unit tests for arguments, reproducibility (with `random_state`), and some testable mathematical properties.
- [x] All tests pass locally.
- [x] All public functions have docstrings and examples.
- [x] Code style follows existing content in `_multivariate.py`.
- [x] Docstrings with new functionality.
- [x] Added myself to `THANKS.txt`
- [x] Added a documentation page via the docstring. 

**Would appreciate feedback on:**

1. How to unit test `rvs`?
1. Should we support `cdf` and `logcdf` given that the multivariate t-distribution does not have a closed-form solution for the CDF? While I have seen some implementations (see https://github.com/scipy/scipy/issues/10042#issuecomment-576806054 in particular), I don't yet fully understand how they work and would prefer adding these methods to be a separate PR.
1. Should we support `entropy`? I don't know anything about the differential entropy of the multivariate t-distribution, and would also prefer punting that to a future PR.

gwgundersen commented:
My tests pass locally, but the CI tests fail for a reason I don't understand, e.g.:

```
loc = rng.random(3)
E   AttributeError: 'mtrand.RandomState' object has no attribute 'random'
```

mdhaber commented:
@gwgundersen Thanks for submitting this. Other opinions may differ, but I suspect it's OK to start without `cdf` and `entropy` in this PR and add later, since not all multivariate distributions have all the methods we might like. 

I'm not sure if this will help, but try `git submodule update --init` before `make html`. You can also see the documentation built by CI: click "Details".
![image](https://user-images.githubusercontent.com/6570539/89163895-2dfbad00-d52b-11ea-8419-6300d942bca4.png)

It looks like it's not showing up in the documentation. I think you need to add your distribution to `scipy/stats/__init__.py`:
```
Multivariate distributions
==========================
.. autosummary::
   :toctree: generated/
   multivariate_normal   -- Multivariate normal distribution
   matrix_normal         -- Matrix normal distribution
   dirichlet             -- Dirichlet
   wishart               -- Wishart
   invwishart            -- Inverse Wishart
   multinomial           -- Multinomial distribution
   special_ortho_group   -- SO(N) group
   ortho_group           -- O(N) group
   unitary_group         -- U(N) group
   random_correlation    -- random correlation matrices
```

I'm not sure about the failing test or how to test the RVS; perhaps @WarrenWeckesser has thoughts?

gwgundersen commented:
@mdhaber, thanks for that suggestion. I tried initializing and updating the git submodule, but it did not seem to work. Is it possible I need to rebuild Scipy from scratch to update the version number based on the commit? I've tried to do that and, oddly, am completely failing to do so—I was able to build Scipy previously, and I don't know if it's a change in the build procedure or something local to me. I wrote more details [here](https://github.com/scipy/scipy/issues/10042#issuecomment-674455994).

mdhaber commented:
@gwgundersen yes, it sounds to me like you need to build SciPy again, too. I'm not sure how to fix your build error, though. Sometimes my approach has been to nuke my environment and start from scratch, but I see that you've written up your issue, so hopefully you'll get a more helpful reply. I've also seen [this](https://github.com/scipy/scipy/issues/12489) issue for macs lately, but it looks different from yours.

rlucas7 commented on THANKS.txt:
```diff
@@ -240,6 +240,7 @@ Luigi F. Cruz for adding time/frequency domain option to scipy.signal.resample.
 Wesley Alves for improvements to scipy.stats.jarque_bera and scipy.stats.shapiro
 Mark Borgerding for contributing linalg.convolution_matrix.
 Shashaank N for contributions to scipy.signal.
+Gregory Gundersen for implementing the multivariate t-distribution.
```
Can this line be removed from the PR? 
See [thread](https://mail.python.org/pipermail/scipy-dev/2020-August/024331.html) for discussion (changes to this file causes merge conflicts adding effort to contributors and maintainers). 
For now, I've opened a [wiki](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.6.0*)
Once the PR is merged, we add thanks to the wiki. Similar process to what we followed in 1.5.0 release.

gwgundersen replied:
Okay, I've removed that line from `THANKS.txt` and added it to the wiki instead.

mdhaber commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,368 @@ def rvs(self, dim, size=1, random_state=None):
+
+class multivariate_t_gen(multi_rv_generic):
+    r"""
+    A multivariate t-distributed random variable.
```
```suggestion
    A multivariate t-distributed random variable.
    
```
That should fix the rendering on the top level stats documentation page:
![image](https://user-images.githubusercontent.com/6570539/92642386-f522c680-f294-11ea-9ba3-780015e238b3.png)

mdhaber commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,368 @@ def rvs(self, dim, size=1, random_state=None):
+    In addition to calling the methods below, the object itself may be called
+    as a function to fix the location, shape matrix, and degrees of freedom
+    parameters, returning a "frozen" multivariate t-distribution random.
+    .. versionadded:: 1.5.0
```
This isn't rendering properly ([here](https://21176-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.stats.multivariate_normal.html#scipy.stats.multivariate_normal) is what it looks like currently). Please take a look at the source for some of the other distributions and add whitespace as they do. Unfortunately, Sphinx is very particular about that sort of thing.

gwgundersen replied:
Can we resolve this? The docs render properly for me, I assume because of @rlucas7's changes [here](https://github.com/scipy/scipy/pull/12582#issuecomment-691762038).

rlucas7 replied:
> Can we resolve this? The docs render properly for me, I assume because of @rlucas7's changes [here](https://github.com/scipy/scipy/pull/12582#issuecomment-691762038).

Sometimes it takes github (browser application) a little while to update the client rendering with diff changes. I wasn't seeing the "outdated" note before, see it now though and agree the changes are taken care of, marking resolved.

rlucas7 commented:
@mdhaber and @gwgundersen I did some testing of this PR locally to see if I could get the docstrings to build. 

I couldn't set my local changes to track this pr for some reason (not sure why) so I posted the changes here:

https://github.com/rlucas7/scipy/commit/56c9d2738b0038af6f4a9ab1d46976aa8fab574d

it is mostly spacing around sections of docstrings (e.g. before example line, etc). 
The examples render locally for my html and latex doc builds. 
There may be an extra newline character here or there but that should give you an idea of what's left for docstring changes. 

Let me know if I can help further to get this PR to merging, seems close and just needs a few more small changes.

mdhaber commented:
@rlucas7 Do you want me to add those to this PR?

gwgundersen commented:
@mdhaber and @rlucas7, thanks for keeping this fire alive. I should have time to look at this again later in the week. Also, I have access to a new machine, which may help with my issues building the docs. I hope that building the docs correctly and pushing will fix some of the failing tests.

rlucas7 commented:
@mdhaber wrote: 
> @rlucas7 Do you want me to add those to this PR?

If @gwgundersen is ok with that then yes. 
Would be nice to get the doc build to green so we can focus on the scientific/probabilistic aspects of the PR and move this one forward.

rlucas7 commented:
> @mdhaber and @rlucas7, thanks for keeping this fire alive. I should have time to look at this again later in the week. Also, I have access to a new machine, which may help with my issues building the docs. I hope that building the docs correctly and pushing will fix some of the failing tests.

No worries @gwgundersen please let us know if you run into issues w/dev setup.

mdhaber commented:
OK I went ahead and pushed @rlucas7's changes to @gwgundersen's branch so we can see how they affect the doc rendering.
@gwgundersen you'll need to pull the changes e.g. `git pull origin/10042-multivariate-t-again` before continuing work.

mdhaber commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,405 @@ def rvs(self, dim, size=1, random_state=None):
+    >>> x, y = np.mgrid[-1:1:.01, -1:1:.01]
+    >>> pos = np.dstack((x, y))
+    >>> rv = multivariate_t([1.0, -0.5], [[2.1, 0.3], [0.3, 1.5]], df=2)
+    >>> plt.contourf(x, y, rv.pdf(pos))
```
This renders as:
![image](https://user-images.githubusercontent.com/6570539/93381445-88b84200-f815-11ea-9429-7010b02c5b00.png)

Might want to make the aspect ratio square and choose a more complete window.

gwgundersen replied:
Okay, I made the window more complete.

To my knowledge, changing the aspect ratio would require making an axes and then setting it, which would add a lot of boilerplate to the small example IMO. I just followed the contour plot [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.multivariate_normal.html#scipy.stats.multivariate_normal).

gwgundersen replied:
@mdhaber, thoughts? Here is the current version with a more complete window:

![](https://user-images.githubusercontent.com/2818208/94142371-2cb87380-fe3c-11ea-9dd7-06eda57004a4.png)

To make the aspect ratio square, we would need to add two lines to the example:

```python
>>> import numpy as np
>>> import matplotlib.pyplot as plt
>>> from scipy.stats import multivariate_t
>>> x, y = np.mgrid[-1:3:.01, -2:1.5:.01]
>>> pos = np.dstack((x, y))
>>> rv = multivariate_t([1.0, -0.5], [[2.1, 0.3], [0.3, 1.5]], df=2)
>>> fig, ax = plt.subplots(1, 1)                                            # This line
>>> ax.set_aspect('equal')                                                  # And this line
>>> plt.contourf(x, y, rv.pdf(pos))
```

mdhaber replied:
I thought it would be nice to keep it square so that the effect of the shape parameter is visually clear.
I am OK with adding those two lines to do that. 
@rlucas7 is it worth it?

rlucas7 replied:
> I thought it would be nice to keep it square so that the effect of the shape parameter is visually clear.
> I am OK with adding those two lines to do that.
> @rlucas7 is it worth it?

I do think that visually getting a full equi-contour trace is helpful. 
Especially for users who are less familiar w/the distribution. 
The rendering will be helpful to refer to if we get questions and upshot is the figure stays within line char limit. 
We can omit the two end of line comments:

```
>>> fig, ax = plt.subplots(1, 1)                                            # This line
>>> ax.set_aspect('equal')                                                  # And this line
```
and keep the rest of the two added lines.
Unless anyone has major objections, let's use the manual axis.

gwgundersen replied:
Okay, I've added those two lines.

rlucas7 replied:
ok, I'm marking as resolved @mdhaber if you had other concerns on this part please reopen.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+
+    def test_reproducibility(self):
+        rng = np.random.RandomState(4)
+        loc = rng.random(3)
```
This line is causing the CI build failures: 

https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=8377&view=ms.vss-test-web.build-test-results-tab&runId=1060078&resultId=140928&paneView=debug

mdhaber replied:
I think all the `rng`s have `uniform`.

rlucas7 replied:
I think I've got changes to get this test to pass. I'll post the suggestings here in the PR and @mdhaber if you agree feel free to commit to re-run CI. The changes I've got passed a local build of this PR, I'm running: 

```
>>> np.__version__
'1.19.1'
>>> scipy.__version__
'1.6.0.dev0+90d7695'
```

mdhaber commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+
+    def test_reproducibility(self):
+        rng = np.random.RandomState(4)
+        loc = rng.random(3)
```
```suggestion
        loc = rng.uniform(3)
```

rlucas7 replied:
@mdhaber it looks like the only way I could change this line via a suggestion was to backout your previous line change and edit the suggestion with a new one. The suggested change renders as you suggesting and not yet commited and does not allow the commit button to be changed-very weird. 

Are you able to (re-)commit this suggested changed? `rng.uniform(0, 1, 3)` worked for me to get this test to pass.

rlucas7 replied:
this suggested change was committed already, closing. Thanks @mdhaber

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+        rng = np.random.RandomState(4)
+        loc = rng.random(3)
+        tmp = rng.random((20, 3))
+        shape = np.matmul(tmp.T, tmp)
```
I'd be in favor of either using an explicit covariance matrix here or drawing from a Wishart, former is likely easier. See suggested change beneath.

mdhaber replied:
I committed it; now `shape` is an identity matrix.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+        shape = np.eye(4)
+        df = 1
+
+        try:
+            multivariate_t(loc, shape, df, allow_singular=False)
+        except np.linalg.LinAlgError:
+            pytest.fail("Test failed by unexpectedly raising `LinAlgError`.")
```
Not sure this is what you want to do here. Is the purpose of this test to test that a `LinAlgError` is raised? 

If so, then usually constructing a matrix explicitly is the preferred way to test the exception raising.

gwgundersen replied:
Yeah, the first `try/catch` was a mistake. I think this version of the test is pretty clear:

```python
def test_allow_singular(self):
    loc = np.zeros(4)
    shape = np.eye(4)
    df = 1

    # Make shape singular and verify error was raised.
    shape[2, 2] = 0
    allow_singular = False
    seed = 0
    assert_raises(np.linalg.LinAlgError,
                  multivariate_t,
                  loc, shape, df, allow_singular, seed)

    # Check that singular `shape` is allowed when `allow_singular=True`.
    try:
        multivariate_t(loc, shape, df, allow_singular=True)
    except np.linalg.LinAlgError:
        pytest.fail("Test failed by unexpectedly raising `LinAlgError`.")
```

I don't like this test:

```python
val = ss.multivariate_t(loc, shape, df, allow_singular=False).pdf(loc)
assert_almost_equal(val, 0.07599088773175335)
```

because the thing we're asserting is a side-effect of the correct behavior we want to test, namely that `LinAlgError` wasn't raised. Thoughts?

gwgundersen replied:
Yeah, I see your point. The `_PSD` class is also well-tested in the `multivariate_normal` section. I've removed the second test, so the full unit test is just:

```python
def test_allow_singular(self):
    # Make shape singular and verify error was raised.
    args = dict(loc=[0,0], shape=[[0,0],[0,1]], df=1, allow_singular=False)
    assert_raises(np.linalg.LinAlgError, multivariate_t, **args)
```
The main check of the original unit test is that the `allow_singular` argument is correctly passed through in `__call__`, which is perhaps a minor concern. With this new test, if you don't pass the argument through, all tests still pass.

rlucas7 replied:
> Yeah, I see your point. The `_PSD` class is also well-tested in the `multivariate_normal` section. I've removed the second test, so the full unit test is just:
> 
> ```python
> def test_allow_singular(self):
>     # Make shape singular and verify error was raised.
>     args = dict(loc=[0,0], shape=[[0,0],[0,1]], df=1, allow_singular=False)
>     assert_raises(np.linalg.LinAlgError, multivariate_t, **args)
> ```
> 
> The main check of the original unit test is that the `allow_singular` argument is correctly passed through in `__call__`, which is perhaps a minor concern. With this new test, if you don't pass the argument through, all tests still pass.

Ok looks good-thanks.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+                      loc, shape, df, allow_singular, seed)
+
+        # Check that singular `shape` is when when `allow_singular=True`.
+        try:
+            multivariate_t(loc, shape, df, allow_singular=True)
+        except np.linalg.LinAlgError:
+            pytest.fail("Test failed by unexpectedly raising `LinAlgError`.")
+
```
similar as above.

gwgundersen replied:
Please see [this comment](https://github.com/scipy/scipy/pull/12582/files/b16030a2e5c423231106c15ca5bc093e475a0c32#r493549941).

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+    def test_reproducibility(self):
+        rng = np.random.RandomState(4)
+        loc = rng.random(3)
+        tmp = rng.random((20, 3))
+        shape = np.matmul(tmp.T, tmp)
```
```suggestion
        shape = np.eye(3)
```

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+            pytest.fail("Test failed by unexpectedly raising `LinAlgError`.")
+
+        # Make shape singular and verify error was raised.
+        shape[2, 2] = 0
```
so this part seems like exhibiting an explicit singular matrix, I'm confused on this unit test.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+        P_VAL_MIN = 0.1
+
+        dist = multivariate_t(0, 1, df=100000, seed=1)
+        samples = dist.rvs(size=100000)
```
I'm just wondering what runtime of this one would be, maybe one of the 3 cases is sufficient?

gwgundersen replied:
This is what I get using IPython's `%timeit`:

```
In [1]: from scipy.stats import multivariate_t
In [2]: %timeit dist = multivariate_t(0, 1, df=100000, seed=1); samples = dist.rvs(size=100000)
9.15 ms ± 73.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

We probably don't need all three though.

rlucas7 replied:
Ok, if it's that quick we can safely leave as is, we do care about correct and coverage of tests but also are cognizant and make tradeoffs on speed for maintenance reasons as well-these can be left as is but maybe change the name of the test from `*_approx_normal` to `*_normal_or_approx`? 

Maintain test prefix string in place of `*`.

gwgundersen replied:
Okay. Per [this reply](https://github.com/scipy/scipy/pull/12582#discussion_r492029144), I've opted to create two unit tests, `*_approx_normal` and `*_is_normal` for when `df` is large and `np.inf` respectively.

rlucas7 replied:
marking resolved.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+        _, p = normaltest(samples)
+        assert ((p > P_VAL_MIN).all())
+
+        dist = multivariate_t(0, 1, df=np.inf, seed=7)
```
for `df=np.inf` the returned results should be ***exactly*** normal.

gwgundersen replied:
Seems related to [this comment](https://github.com/scipy/scipy/pull/12582#discussion_r489905679).

Does this mean adding an explicit check for `df == np.inf` and returning a `multivariate_normal` instance instead, e.g.:

```python
def __call__(self, loc=None, shape=1, df=1, allow_singular=False,
             seed=None):
    """
    Create a frozen multivariate t-distribution. See
    `multivariate_t_frozen` for parameters.

    """
    if df == np.inf:
        return multivariate_normal_frozen(loc=loc, cov=shape, 
                                          allow_singular=allow_singular, seed=seed)
    return multivariate_t_frozen(loc=loc, shape=shape, df=df,
                                 allow_singular=allow_singular, seed=seed)
```

And then adding docs / tests for this?

rlucas7 replied:
> Seems related to [this comment](https://github.com/scipy/scipy/pull/12582#discussion_r489905679).
> 
> Does this mean adding an explicit check for `df == np.inf` and returning a `multivariate_normal` instance instead, e.g.:
> 
> ```python
> def __call__(self, loc=None, shape=1, df=1, allow_singular=False,
>              seed=None):
>     """
>     Create a frozen multivariate t-distribution. See
>     `multivariate_t_frozen` for parameters.
> 
>     """
>     if df == np.inf:
>         return multivariate_normal_frozen(loc=loc, cov=shape, 
>                                           allow_singular=allow_singular, seed=seed)
>     return multivariate_t_frozen(loc=loc, shape=shape, df=df,
>                                  allow_singular=allow_singular, seed=seed)
> ```

Yes, both here in `__call__` method and maybe in `_rvs` too. Now that I've looked over the PR again it looks like he `pdf` and `cdf`, `logcdf` methods might need to have branches as well. 

> And then adding docs / tests for this?

I think you already have a test here but for other methods where you add a branch you'd want to add a test case.

gwgundersen replied:
Both the `pdf` and `logpdf` methods delegate to `multivariate_t._logpdf`, which allowed me to place a second check for `df == np.inf` there. This required me to add a `rank` argument to `_logpdf`, which was already being computed by the `_PSD` helper class.

Additionally, I've created a new unit test checking for exact normality when `df == np.inf`. Since `multivariate_normal` is well-tested, I just handle two cases:

1. When `__call__` is called, I confirm the returned object is `multivariate_normal_frozen`.
2. When `pdf` or `logpdf` are called, I verify `multivariate_normal._logpdf` was called via mocking.

This PR doesn't support `cdf` or `logcdf`.

rlucas7 replied:
> I don't think I need to check for `df == np.inf` for the `rvs` method since the code is effectively:
> 
> ```python
> # when df == np.inf:
> x = np.ones(size)
> z = rng.multivariate_normal(np.zeros(dim), shape, size=size)
> samples = loc + z / np.sqrt(x)[:, None]
> ```

Yes I see that now, should be fine as written. 

> This PR doesn't support `cdf` or `logcdf`.

Got it, was confusing with univariate distributions for some reason. No need for `cdf` or `logcdf` tests here.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+    @pytest.mark.parametrize("x, mu, df, shape, ans", MULTIVARIATE_T_TESTS)
+    def test_logpdf_correct(self, x, mu, df, shape, ans):
+        dist = multivariate_t(mu, df, shape, seed=0)
+        # We already test this elsewhere, but let's explicitly confirm the
+        # PDF is correct since this test requires checking the log of that
+        # value.
+        val1 = dist.pdf(x)
+        assert_array_almost_equal(val1, ans)
```
I think this value is already tested in the test `test_pdf_correctness` correct? 

If so no need to test again.

gwgundersen replied:
Okay, removed this duplicate assert.

rlucas7 replied:
marking resolved, thanks.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,405 @@ def rvs(self, dim, size=1, random_state=None):
+            else:
+                dim = shape.shape[0]
+            loc = np.zeros(dim)
+        else:
```
IIUC this branch could be entered if `loc is not None and shape is None` is one more `elif` branch needed?

gwgundersen replied:
No, since `shape` defaults to `1`. Therefore `multivariate_t(loc=val)` takes the `else` branch.

rlucas7 replied:
Ok, but what if a `None` is passed in explicitly as an argument?

gwgundersen replied:
Okay, I see. I also wasn't handling the case of `loc=None` correctly. I amended the behavior to mimic `multivariate_normal`, e.g.:

```
>>> multivariate_t(loc=[1,-1], shape=None).shape
array([[1., 0.],
       [0., 1.]])
```

and added [8 tests for all permutations of `None` for the three arguments](https://github.com/scipy/scipy/pull/12582/files#diff-edd4ec9cd7afebecda51f7f29124af70R1765).

rlucas7 replied:
see it now, marking resolved.

rlucas7 reviewed (changes requested):
@mdhaber I left a couple suggested changes to get the `rvs` reproducibility test to pass for me locally. One had some weird issue trying to overwrite the suggested change you made, perhaps you can take a look, and if they look ok to you, commit and see if gets to green? 

I also left several comments, mostly minor,  2 less minor ones though.

mdhaber commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1638,187 @@ def test_haar(self):
+
+    def test_reproducibility(self):
+        rng = np.random.RandomState(4)
+        loc = rng.uniform(3)
```
```suggestion
        loc = rng.uniform(size=3)
```

gwgundersen commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,405 @@ def rvs(self, dim, size=1, random_state=None):
+    Notes
+    -----
+    %(_mvt_doc_callparams_note)s
+    The matrix `shape` must be a (symmetric) positive definite matrix. The
```
Yes, you're correct. PSD is what is used in the docstrings for both the `multivariate_normal` code and `_PSD` utility class. I've changed references from "positive definite" to "positive semidefinite".

rlucas7 replied:
yes I see the changes, marking resolved.

gwgundersen commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,405 @@ def rvs(self, dim, size=1, random_state=None):
+
+        """
+        dev  = x - loc
+        maha = np.square(np.dot(dev, prec_U)).sum(axis=-1)
```
Yes, Mahalanobius distance. I just copied the variable names from [here](https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py#L474). Happy to add a comment if we think it's unclear.

rlucas7 replied:
> Yes, Mahalanobius distance. I just copied the variable names from [here](https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py#L474). Happy to add a comment if we think it's unclear.

Oh, I hadn't noticed that, internal consistency is fine so no need to change-please ignore the comment.

gwgundersen commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,405 @@ def rvs(self, dim, size=1, random_state=None):
+shape : array_like, optional
+    Positive definite matrix of the distribution. (default ``1``)
+df : integer, optional
+    Degrees of freedom of the distribution; must be a positive whole number. (default ``1``) 
```
Are you suggesting two changes:

1. If the user provides a non-integer `df` value, defer to a gamma mixture representation?
2. If the user provides `df = np.inf`, defer to a multivariate normal instance?

I haven't heard of the gamma mixture representation, but I can look into it if you have a reference.

rlucas7 replied:
> Are you suggesting two changes:
> 
>     1. If the user provides a non-integer `df` value, defer to a gamma mixture representation?

Yes, I think this only applies for the `_rvs` method though, please double check but I think the `pdf` etc. methods are ok. The chi-squared distributed is just a special case of gamma. Using wikipedia notation for gamma and chi-squared distributions the shape, `k` here: 
https://en.wikipedia.org/wiki/Gamma_distribution
corresponds to `df/2` and rate is `2`. 
You can use this for all positive values and the `df/2` correspondence will work. 
Think of it like an analytic continuation of the degrees of freedom to positive reals. 

>     2. If the user provides `df = np.inf`, defer to a multivariate normal instance?

Yes.

rlucas7 replied:
@gwgundersen I looked at the `chi2` implementation a little closer today and confirmed that the method ***will*** take non-integer values for the `df` parameter. I think this is more a matter of documentation change and maybe adding a test or 2 to confirm that non-integral values are accepted. Apologies for the confusion.

rlucas7 reviewed (changes requested):
@gwgundersen hopefully my review/clarifying comments here help. Ping me in comments if still unclear. I think is mostly a matter of adding some branching for `np.inf` df case.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,407 @@ def rvs(self, dim, size=1, random_state=None):
+        dev  = x - loc
+        maha = np.square(np.dot(dev, prec_U)).sum(axis=-1)
+
+        t = 0.5 * (df + dim)
+        A = gammaln(t)
+        B = gammaln(0.5 * df)
+        C = dim/2. * np.log(df * np.pi)
+        D = 0.5 * log_pdet
+        E = -t * np.log(1 + (1./df) * maha)
```
For example for any non-negative value of `df` thee calculations here should work fine. So I think `int` is not a strict requirement for the `df` arg, `df>0` is a requirement though.

rlucas7 replied:
marking resolved

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,407 @@ def rvs(self, dim, size=1, random_state=None):
+        # Process degrees of freedom.
+        if df is None:
+            df = 1
+        if not isinstance(df, int) and not np.isinf(df):
```
so basically this is the only line to change on the `df` comemtns. It looks like all other instances come into this one. I would make this a check on `df>0` and swap out the `np.inf` check with a (Nan or None) check.

gwgundersen replied:
I've added branches for `df <= 0` and `np.isnan(df)`, raising a `ValueError` in both cases. I've also added a unit test to confirm we raise/don't raise errors for various arguments.

gwgundersen replied:
Also amended the docs/docstrings.

rlucas7 replied:
great, thanks for that, marking resolved.

rlucas7 reviewed (commented):
@gwgundersen apologies for my confusing review yesterday. For the `df` parameter the way things are currently written in this PR, everything should work correctly if `df` is e.g. `1.1` except for the check that you have for `int`. So replace that check with the suggested change and I think the PR is good from that side. I think my comments for when `df==np.inf` is/are still valid though, let me know if unclear.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1643,248 @@ def test_haar(self):
+
+    def test_scalar_list_and_ndarray_arguments(self):
+        dist = multivariate_t(-1, 2, 3)
+        assert(dist.loc == np.array([-1]))
+        assert(dist.shape == np.array([2]))
+        assert(dist.df == 3)
```
suggest to swap `assert` with `assert_{allclose,equal}` here and throughout for consistency.

gwgundersen replied:
Okay, I'm now using `assert_equal`. This catches some small discrepancies. For example, this passes an `assert`:

```python
dist.shape == np.array([2])
# array([[ True]])
```

but not an `assert_equal`. The real value should be `[[2]]` or `np.array([[2]])`.

I've also wrapped the test using `parameterize`.

rlucas7 replied:
weird, must be a consistency issue w/the github UI, I now see changes, marking resolved, apologies for noise.

rlucas7 commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1643,248 @@ def test_haar(self):
+
+    def test_argument_error_handling(self):
+        # `loc` should be a one-dimensional vector.
+        with pytest.raises(ValueError):
+            multivariate_t(loc=[[1, 1]])
+
+        # `shape` should be scalar or square matrix.
+        with pytest.raises(ValueError):
+            multivariate_t(shape=[[1, 1], [2, 2], [3, 3]])
+
+        # Confirm df > 0, otherwise `ValueError`:
+        with pytest.raises(ValueError):
+            multivariate_t(df=-1)
+        with pytest.raises(ValueError):
+            multivariate_t(df=0)
```
these could be `assert_raises`, would be less code and context managers do not add much here.

gwgundersen replied:
Sure, I've switched to `assert_raises`. I suppose this fits the paradigm in this file, but FWIW both [this StackOverflow](https://stackoverflow.com/questions/23337471/) and the [Pytest docs](https://docs.pytest.org/en/stable/assert.html#assertraises) suggest the context manager approach. I prefer it because you see the function call directly, rather than implicitly, e.g.:

```python
assert_raises(ValueError,
              multivariate_t,
              **dict(loc=loc, shape=shape, df=df))
```

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,424 @@ def rvs(self, dim, size=1, random_state=None):
+shape : array_like, optional
+    Positive semidefinite matrix of the distribution. (default ``1``)
+df : float, optional
+    Degrees of freedom of the distribution; must be greater than zero or ``np.inf``. (default ``1``) 
```
```suggestion
    Degrees of freedom of the distribution; must be greater than zero. If ``np.inf`` then results are multivariate normal. The default is ``1``. 
```

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,424 @@ def rvs(self, dim, size=1, random_state=None):
+        else:
+            rng = self._random_state
+
+        if df == np.inf:
```
```suggestion
        if np.isinf(df):
```

rlucas7 reviewed (changes requested):
@gwgundersen left a couple more comments. I noticed something weird with the shapes, not sure the cause but behavior is inconsistent:

```
>>> import numpy as np
>>> import scipy.stats as ss
>>> loc = np.zeros(4)
>>> shape = np.eye(4)
>>> df = 1
>>> shape[2, 2] = 0
>>> ss.multivariate_t(loc, shape, df, allow_singular=True).pdf(np.zeros(4))
array([0.07599089])
>>> ss.multivariate_t(loc, shape, np.inf, allow_singular=True).pdf(np.zeros(4))
0.06349363593424098 
```
Could you take a look and add a test for that one?  

Left a couple suggestions for the last unit test that was previously puzzling me-now I think I understand that one better. 

Most other comments are cosmetic or style changes.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3860,7 +3860,7 @@ def rvs(self, dim, size=1, random_state=None):
     Positive semidefinite matrix of the distribution. (default ``1``)
 df : float, optional
-    Degrees of freedom of the distribution; must be greater than zero or ``np.inf``. (default ``1``) 
+    Degrees of freedom of the distribution; must be greater than zero. If ``np.inf`` then results are multivariate normal. The default is ``1``. 
```
hmm, these should've been on new lines. 
I think we have a char limit check in CI of 79 that fails builds.

gwgundersen replied:
Okay, fixed.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3860,7 +3860,7 @@ def rvs(self, dim, size=1, random_state=None):
     Positive semidefinite matrix of the distribution. (default ``1``)
 df : float, optional
-    Degrees of freedom of the distribution; must be greater than zero or ``np.inf``. (default ``1``) 
+    Degrees of freedom of the distribution; must be greater than zero. If ``np.inf`` then results are multivariate normal. The default is ``1``. 
```
```suggestion
    Degrees of freedom of the distribution; must be greater than zero. 
    If ``np.inf`` then results are multivariate normal. The default is ``1``. 
```

gwgundersen replied:
Just fixed locally and pushed, so ignoring.

gwgundersen commented:
> I noticed something weird with the shapes, not sure the cause but behavior is inconsistent:
> 
> ```
> >>> import numpy as np
> >>> import scipy.stats as ss
> >>> loc = np.zeros(4)
> >>> shape = np.eye(4)
> >>> df = 1
> >>> shape[2, 2] = 0
> >>> ss.multivariate_t(loc, shape, df, allow_singular=True).pdf(np.zeros(4))
> array([0.07599089])
> >>> ss.multivariate_t(loc, shape, np.inf, allow_singular=True).pdf(np.zeros(4))
> 0.06349363593424098 
> ```

Ah, good catch. When using `np.inf`, we delegate to `multivariate_normal`, which uses `_squeeze_output`, converting values like `[[2]]` to `2`. I've added `_squeeze_output` to `multivariate_t`'s `_logpdf` function and `rvs` functions. I've added a new unit test `test_shape_correctness`:

```
def test_shape_correctness(self):
    # pdf and logpdf should return scalar when the 
    # number of samples in x is one.
    dim = 4
    loc = np.zeros(dim)
    shape = np.eye(dim)
    df = 1
    x = np.zeros(dim)
    res = multivariate_t(loc, shape, df).pdf(x)
    assert np.isscalar(res)
    res = multivariate_t(loc, shape, df).logpdf(x)
    assert np.isscalar(res)

    # pdf() and logpdf() should return probabilities of shape 
    # (n_samples,) when x has n_samples.
    n_samples = 7
    x = np.random.random((n_samples, dim))
    res = multivariate_t(loc, shape, df).pdf(x)
    assert (res.shape == (n_samples,))
    res = multivariate_t(loc, shape, df).logpdf(x)
    assert (res.shape == (n_samples,))

    # rvs() should return scalar unless a size argument is applied.
    res = multivariate_t(np.zeros(1), np.eye(1), 1).rvs()
    assert np.isscalar(res)

    # rvs() should return vector of shape (size,) if size argument 
    # is applied.
    size = 7
    res = multivariate_t(1, 1, 1).rvs(size=size)
    assert (res.shape == (size,))
```

gwgundersen commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1643,248 @@ def test_haar(self):
+        dist = multivariate_t(loc=[7, 7], shape=[[7, 0], [0, 7]], df=7)
+        assert_equal(dist.loc, [7, 7])
+        assert_equal(dist.shape, [[7, 0], [0, 7]])
+        assert (dist.df == 7)
```
Good idea. This is done.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3943,6 +3943,7 @@ class multivariate_t_gen(multi_rv_generic):
 
     Examples
     --------
+    >>> import numpy as np
```
last I knew when the docs build they have `numpy` already setup by default. If you're running the module in your machine as a stand alone file you would need to add that but as far as I recall we haven't changed so safe to remove this line.

mdhaber replied:
Even if we can, should we? Why remove lines that are needed to run example code?

tylerjereddy replied:
We can remove the import--the debate about copy-paste viability is perhaps another matter, but I'm pretty sure the convention in PR reviews has been to discourage import `numpy as np` over and over.

gwgundersen replied:
Okay, I've removed the `numpy` import.

gwgundersen commented:
@rlucas7 and @mdhaber, thanks for all your help and feedback. I _think_ I've addressed all your comments, but please let me know if I missed anything.

gwgundersen commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -1635,6 +1643,243 @@ def test_haar(self):
+
+        # Floats are okay:
+        try:
+            multivariate_t(df=5.0)
```
@rlucas7, should I remove this test for correct behavior when `df` is a real number?

rlucas7 replied:
> @rlucas7, should I remove this test for correct behavior when `df` is a real number?

I think so yes. If there is not an existing test case which contains a `df` arg with a float value, please make a change there as well, so you've got good coverage.

gwgundersen replied:
Okay, I've removed this `try/catch` and used a `df` float value for another test so we have the same coverage.

gwgundersen replied:
Hm, I don't know why the tests aren't passing now after pushing these changes. All the `stats` tests pass locally, but `scipy/linalg/tests/test_decomp_cossin.py` doesn't.

rlucas7 replied:
Hmm, not sure, seems like a pytest configuration error is causing failures. That wasn’t showing up before so guessing it’s random. I’ll try closing and reopening the pr to restart the CI and see if the issue goes away. Fingers crossed.

rlucas7 reviewed (changes requested):
I think just that last change,  @gwgundersen if you can make that change, should be good from my perspective

gwgundersen commented:
@rlucas7, okay done. Thanks again.

rlucas7 commented:
@gwgundersen it looks like there are some changes with checks on an upgraded version of numpy to the failing CI builds. It looks like these are failing on other branches too and they don’t appear related to this PR. Let’s hold off for now and I’ll check back in a couple days (tomorrow is a busy work day).

rlucas7 commented:
@gwgundersen it looks like @WarrenWeckesser put a fix (now in master) that should make these failing builds pass:
https://github.com/scipy/scipy/pull/12902
do you feel up for a rebase of this branch on master? 
If not I can do it over the weekend and ping you to confirm after I force push to the branch.

Either way is fine w/me but wanted to see which you prefer.

gwgundersen commented:
@rlucas7, I think it's safer if you do the rebase. I should learn, but maybe not here. Thanks for asking.

rlucas7 commented:
> @rlucas7, I think it's safer if you do the rebase. I should learn, but maybe not here. Thanks for asking.

I am not able to push to the branch from the git terminal client because the PR is under your repository. The authentication mechanism is different for the git hub web browser than git itself. I should've known this, git is not github. Anyway, you'll need to perform the rebase. 

 If you don't have upstream setup on your local repo already, do this:
https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/configuring-a-remote-for-a-fork
then assuming that is setup on your local copy, try following these instructions: 
https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
note you will run into 2 rebase conflicts, both are because of `THANKS.txt` conflicts. 

Three things worth noting: 
(1) Be sure your master branch is up to date with the upstream master ***before*** doing the rebase. 
(2) After you've completed the rebase you will see a message like: 
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ git status 
On branch test
Your branch and 'origin/10042-multivariate-t-again' have diverged,
and have 195 and 39 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
``` 
this is expected, and you ***do not*** want to perform a git pull. At that point confirm your local branch builds and passes tests by `python runtests.py -v` and assuming everything passes, then do a `git push -f` assuming you have the `10042-multivariate-t-again` tracking the origin remote branch.

(3) Before you begin the rebase process create a local branch that is checked out on the HEAD of the existing state of `10042-multivariate-t-again` this way you have a local copy if something breaks. 

In the worst case, you bring in a bunch of other commits and need to reopen another new pull request.

rlucas7 commented:
Also, @gwgundersen when I rebased your copy locally and reran the tests everything passes.

gwgundersen commented:
Thanks for the detailed instructions. Should I be rebasing from a specific stable commit on `master`? I ask because after following them, my tests do _not_ all pass. Specifically, I see:

```
=================================================== short test summary info ====================================================
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-40-12-20-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-100-50-50-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-40-12-20-float32] - AssertionError: 
FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-100-50-50-float32] - AssertionError: 
============== 4 failed, 29652 passed, 2030 skipped, 10990 deselected, 76 xfailed, 6 xpassed in 441.03s (0:07:21) ==============
```

I also confirmed that these tests fail on `master`. I even created a new directory from a fresh `git clone ...`, built Scipy, and ran tests; and those tests still fail.

### What I did

1. Set an `upstream` branch and updated my local `master` branch:

```
git remote add upstream git@github.com:scipy/scipy.git
git checkout master
git pull upstream master
```

2. Found the best place to perform the rebase:

```
git merge-base 10042-multivariate-t-again master
<SOME HASH>
```

3. Rebased interactively:

```
git rebase -i <SOME HASH>
```

4. Performed the interactive rebase, squashing some minor commits for clarity.

5. Resolved two merge conflicts in `THANKS.txt`.

6. Re-built Scipy (not sure if this is necessary but seems like it should not hurt):

```
python3 setup.py build_ext --inplace
```

7. Ran all tests:

```
python runtests.py -v
```

rlucas7 commented:
IIUC you still need to complete the rebase by doing 

git rebase -i master

On your branch. 
The merge-base gives a MRCA if the two branches and rebase on that commit puts your commits on top of the MRCA commit. The rebase command above inserts the commits on master between the MRCA commit and your first commit on your branch. Your branch commits are stacked on top of the tip (HEAD) commit from master (but on your local branch).

Try the command above and once done with rebase see if you still get the test fail. IIRC that test was failing but was fixed on master. So that’s why I think the commits aren’t all the way up to tip. 

> On Oct 4, 2020, at 10:26 PM, Gregory Gundersen <notifications@github.com> wrote:
> 
> ﻿
> Thanks for the detailed instructions. Should I be rebasing from a specific stable commit on master? I ask because after following them, my tests do not all pass. Specifically, I see:
> 
> =================================================== short test summary info ====================================================
> FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-40-12-20-float32] - AssertionError: 
> FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[True-100-50-50-float32] - AssertionError: 
> FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-40-12-20-float32] - AssertionError: 
> FAILED scipy/linalg/tests/test_decomp_cossin.py::test_cossin[False-100-50-50-float32] - AssertionError: 
> ============== 4 failed, 29652 passed, 2030 skipped, 10990 deselected, 76 xfailed, 6 xpassed in 441.03s (0:07:21) ==============
> I also confirmed that these tests fail on master. I even created a new directory from a fresh git clone ..., built Scipy, and ran tests; and those tests still fail.
> 
> What I did
> 
> Set an upstream branch and updated my local master branch:
> git remote add upstream git@github.com:scipy/scipy.git
> git checkout master
> git pull upstream master
> Found the best place to perform the rebase:
> git merge-base 10042-multivariate-t-again master
> <SOME HASH>
> Rebased interactively:
> git rebase -i <SOME HASH>
> Performed the interactive rebase, squashing some minor commits for clarity.
> 
> Resolved two merge conflicts in THANKS.txt.
> 
> Re-built Scipy (not sure if this is necessary but seems like it should not hurt):
> 
> python3 setup.py build_ext --inplace
> Ran all tests:
> python runtests.py -v
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

gwgundersen commented:
I _think_ I understand. Reading [this tutorial](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request), it seems like my rebase let me to clean up my branch's history, but your rebase is what replays my branch on master. That said, even after adding the second rebase (`git rebase -i master`), the unit tests do not all pass.

And again, if I re-clone the Scipy repo, build, and run tests on `master`, the same tests fail.

gwgundersen commented:
@rlucas, what if I just do a `git push --force `with my rebased branch, despite these failing tests? I suspect the issue has to do with my machine or OS (Ubuntu 20.04.1), since:

1. The unit tests pass for you.
1. The code and tests appear to rely on LAPACK.
1. The same unit tests fail when using a newly cloned version of `master`.
1. The same unit tests fail when using my current branch (not rebased).

Alternatively, I could ask the developers at https://github.com/scipy/scipy/pull/11524 for help, since that PR introduced the unit tests that are failing for me.

gwgundersen commented:
Also, the failing tests "smell" like they may be caused by differing mathematical or algorithmic conventions, e.g.:

```
>>> np.isclose(u1_2, u1)
array([[False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False],
       [False,  True, False, ..., False, False, False]])
>>> u1_2[0,0]
0.058479194
>>> u1[0,0]
-0.058478586
>>> u1_2[0,2]
0.040936727
>>> u1[0,2]
-0.040936757
```

rlucas7 commented:
Try 

python runtests.py -v -s stats 

That will run only the stats tests. If that passes for you locally then yeah I think a force push is fine.

Also, can you leave a comment on the other PR with the failing test stack trace? 

If you’re seeing that fail others are or will too. 

I run a Mac box so if this is specific to Ubuntu (Not sure if it is) I can’t help much. 

-Lucas Roberts 

> On Oct 5, 2020, at 5:21 PM, Gregory Gundersen <notifications@github.com> wrote:
> 
> ﻿
> Also, the failing tests "smell" like they may be caused by differing mathematical or algorithmic conventions, e.g.:
> 
> >>> np.isclose(u1_2, u1)
> array([[False,  True, False, ..., False, False, False],
>        [False,  True, False, ..., False, False, False],
>        [False,  True, False, ..., False, False, False],
>        ...,
>        [False, False, False, ..., False, False, False],
>        [False,  True, False, ..., False, False, False],
>        [False,  True, False, ..., False, False, False]])
> >>> u1_2[0,0]
> 0.058479194
> >>> u1[0,0]
> -0.058478586
> >>> u1_2[0,2]
> 0.040936727
> >>> u1[0,2]
> -0.040936757
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ilayn commented:
Just to make sure that this is not a sign problem (since the results are unique up to columns signs), could you please check if this turns out all true

```python
>>> np.allclose(np.abs(u1_2), np.abs(u1), rtol=0., atol=np.finfo(np.float32).eps*100)
```

gwgundersen commented:
@rlucas7, I rebased, confirmed the stats tests pass, and force pushed. All checks pass via GitHub, and I didn't bring in any additional commits.

gwgundersen commented:
@ilayn, that command returns `False`:

```
> /home/gwg/projects/scipy/build/testenv/lib/python3.7/site-packages/scipy/linalg/tests/test_decomp_cossin.py(151)test_cossin_separate()
-> assert_allclose(u1_2, u1, rtol=0., atol=10*np.finfo(dtype_).eps)
(Pdb) np.allclose(np.abs(u1_2), np.abs(u1), rtol=0., atol=np.finfo(np.float32).eps*100)
False
```

Also, despite my `linalg` tests failing locally, the CI tests all pass.

ilayn commented:
That can happen also due to an outdated blas/lapack distro (mkl or openblas etc.) Because apparently not only signs are wrong but the results are also different so it is an actual miscalculation.

rlucas7 commented:
> That can happen also due to an outdated blas/lapack distro (mkl or openblas etc.) Because apparently not only signs are wrong but the results are also different so it is an actual miscalculation.

Thanks for commenting on this @ilayn I hadn't think about BLAS version issues. 

So that we know in case of future questions, it would be great @gwgundersen if you could let us know which BLAS and version you have running.

rlucas7 commented on scipy/stats/_multivariate.py:
```diff
@@ -3850,3 +3851,427 @@ def rvs(self, dim, size=1, random_state=None):
+        elif df <= 0:
+            raise ValueError("'df' must be greater than zero.")
+        elif np.isnan(df):
+            raise ValueError("'df' must be greater than zero or 'np.inf'.")
```
```suggestion
            raise ValueError("'df' is 'nan' but must be greater than zero or 'np.inf'.")
```

rlucas7 reviewed (approved):
making 2 minor (cosmetic) changes to docs and marking approved. I'll leave open until the upcoming weekend to give others some time to comment. If no further comments by Sunday October 11th, at 9am EST, I'll merge to master (sometime after 9am).

gwgundersen commented:
@ilayn, I'm not sure which BLAS and LAPACK versions are being used. This is what I tried when trying to find out:

```python
>>> import numpy.distutils.system_info as sysinfo

>>> sysinfo.get_info('lapack')
{'libraries': ['lapack', 'lapack'],
 'library_dirs': ['/usr/lib/x86_64-linux-gnu'],
 'language': 'f77'}

>>> sysinfo.get_info('blas') 
{'libraries': ['blas', 'blas'],
 'library_dirs': ['/usr/lib/x86_64-linux-gnu'],
 'include_dirs': ['/usr/local/include',
  '/usr/include',
  '/home/gwg/miniconda3/envs/scipy/include'],
 'language': 'c',
 'define_macros': [('HAVE_CBLAS', None)]}
```

When I `ls` in those directories, I see multiple shared object files:

```
/usr/lib/x86_64-linux-gnu/lapack$ ls
liblapack.a  liblapack.so  liblapack.so.3  liblapack.so.3.9.0

/usr/lib/x86_64-linux-gnu/openblas-pthread$ ls
cmake       libblas.so.3  liblapack.so.3         libopenblasp-r0.3.8.so  pkgconfig
libblas.a   liblapack.a   libopenblas.a          libopenblas.so
libblas.so  liblapack.so  libopenblasp-r0.3.8.a  libopenblas.so.0
```

I also tried to re-build Scipy from scratch pointing to the latest object files:

```
export LAPACK=/usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
export BLAS=/usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
python setup.py install
```

But the tests still fail. Happy to try something else if you have advice, but I can't seem to find any info online as to the actual BLAS or LAPACK versions being used.

rlucas7 commented:
Squashed and merged, nice work @gwgundersen

mdhaber commented:
Thanks @gwgundersen, @rlucas!

hp5588 commented:
Hello @gwgundersen , 
I find the code in the following line has a problem with Lint check and cause failure in _continuous-integration/travis-ci/pr_
I guess it's because of space before "=" 
https://github.com/scipy/scipy/blob/882849f81e4eda89008261b691206c5ae7795d26/scipy/stats/_multivariate.py#L4083

it produces the following error
```
36.65s$ pycodestyle scipy benchmarks/benchmarks
scipy/stats/_multivariate.py:4083:12: E221 multiple spaces before operator
1       E221 multiple spaces before operator
The command "pycodestyle scipy benchmarks/benchmarks" exited with 1.
```

rlucas7 commented:
@hp5588 I'm seeing that now, I've got a PR to fix it. 

ref:
https://github.com/scipy/scipy/pull/12938
